auth/localkeys.go

```diff
@@ -0,0 +1,288 @@
+}
+
+// LoadKeysFromFile loads keys from a file on disk.
+func LoadDelegateKeysFromFile(filename string) error {
+	pub, priv, err := LoadKeyPairFromFile(filename)
+	if err != nil {
+		return err
+	}
+
+	updateDelegateKeys(pub, priv)
+	return nil
+}
+
+// CreateOrLoadMasterKeys will load the master keys from disk
+//  If the file does not exist, it will generate new keys and
+//  write them to disk.
+func CreateOrLoadMasterKeys(filename string) error {
+	if _, err := os.Stat(filename); os.IsNotExist(err) {
+		if err = os.MkdirAll(path.Dir(filename), os.ModeDir|0700); err != nil {
+			return err
+		}
+
+		pub, priv, err := GenerateRSAKeyPairPEM(nil)
+		if err != nil {
+			return err
+		}
+		err = DumpPEMKeyPairToFile(filename, pub, priv)
+		if err != nil {
+			return err
+		}
+
+	} else if err != nil {
+		return err
+	}
+
+	publicKey, privateKey, err := LoadKeyPairFromFile(filename)
+	if err != nil {
+		return err
+	}
+
+	masterKeys = MasterKeys{publicKey, privateKey}
+
+	return nil
+}
+
+// DumpPEMKeyPairToFile dumps PEM-encoded public and private keys to a single file
+func DumpPEMKeyPairToFile(filename string, public, private []byte) error {
+	data, err := DumpRSAPEMKeyPair(public, private)
+	if err != nil {
+		return err
+	}
+	return ioutil.WriteFile(filename, data, 0666)
+}
+
+// LoadPEMKeyPair loads a private/public key pair from a reader over PEM-encoded data.
+//  The private key is first, the public key is second.
+func LoadKeyPairFromFile(filename string) (public crypto.PublicKey, private crypto.PrivateKey, err error) {
+	data, err := ioutil.ReadFile(filename)
+	if err != nil {
+		return nil, nil, err
+	}
+	return LoadRSAKeyPairPackage(data)
+}
+
+// LoadDelegateKeysFromPEM loads the local delegate keys (master public, delegate private)
+//  from PEM data passed in directly
+//  Useful mostly for writing tests
+func LoadDelegateKeysFromPEM(public, private []byte) error {
+	pub, priv, err := LoadRSAKeyPair(public, private)
+	if err != nil {
+		return err
+	}
+	updateDelegateKeys(pub, priv)
+	return nil
+}
+
+// LoadMasterKeysFromPEM loads the local master keys from PEM data passed in directly
+//  Useful mostly for writing tests
+func LoadMasterKeysFromPEM(public, private []byte) error {
+	pub, priv, err := LoadRSAKeyPair(public, private)
+	if err != nil {
+		return err
+	}
+
+	masterKeys = MasterKeys{pub, priv}
+	return nil
+}
+
+// ClearKeys wipes the current state
+func ClearKeys() {
+	masterKeys = MasterKeys{}
+	updateDelegateKeys(nil, nil)
+}
+
+// WaitForDelegateKeys blocks until delegate keys are defined.
+func WaitForDelegateKeys() {
+	dKeyCond.L.Lock()
+	defer dKeyCond.L.Unlock()
+	for delegateKeys.localPrivate == nil || delegateKeys.masterPublic == nil {
+		dKeyCond.Wait()
+	}
+}
+
+// WatchDelegateKeyFile watches the delegate key file on the filesystem and
+// updates the internal delegate keys when changes are detected.
+func WatchDelegateKeyFile(filename string, cancel chan interface{}) error {
```

Comment:
We discussed this; it's `interface{}` because the shutdown channel in daemon is `interface{}`